File: union_find/javascript/QuickUnion.js
class QuickUnion {

  constructor(N) {
    var i
    this.id = []
    for(i = 0; i < N; i++) {
      this.id.push(i)
    }
  }

  root(i) {
    while(i != this.id[i]) i = this.id[i]
    return i
  }

  connected(p, q) {
    return this.root(p) == this.root(q)
  }

  union(p, q) {
    var pRoot = this.root(p)
    var qRoot = this.root(q)

    this.id[pRoot] = qRoot
    console.log(this.id)
  }
}

File: union_find/ruby/lib/UF/quick_union.rb
# Array as a tree

class UF::QuickUnion
  attr_reader :id

  def initialize(n)
    @n = n.to_i
    @id = (0..@n-1).to_a
  end

  def union(p, q)
    return if self.connected?(p, q)
    pRoot = self.root(p)
    qRoot = self.root(q)

    @id[pRoot] = qRoot
  end

  def connected?(p, q)
    root(p) == root(q)
  end

  private

    def root(i)
      while(id[i] != i) do
        i = id[i]
      end
      i
    end
end
